src/services/kratos/auth-email-no-password.ts

```diff
@@ -86,21 +230,243 @@ export const AuthWithEmailPasswordlessService = (): IAuthWithEmailPasswordlessSe
       })
       const sessionToken = result.data.session_token as SessionToken
 
-      // note: this only works when whoami: required_aal = aal1
-      const kratosUserId = result.data.session.identity.id as UserId
+      // identity is only defined when identity has not enabled totp
+      const kratosUserId = result.data.session.identity?.id as UserId
 
       return { sessionToken, kratosUserId }
     } catch (err) {
       return new UnknownKratosError(err)
     }
   }
 
+  const addUnverifiedEmailToIdentity = async ({
+    kratosUserId,
+    email,
+  }: {
+    kratosUserId: UserId
+    email: EmailAddress
+  }) => {
+    // TODO: replace with ?
+    // const identity = await IdentityRepository().getIdentity(kratosUserId)
+    // if (identity instanceof Error) return identity
+    // if (identity.schema !== SchemaIdType.PhoneNoPasswordV0) {
+
+    let identity: KratosIdentity
+
+    try {
+      ;({ data: identity } = await kratosAdmin.getIdentity({ id: kratosUserId }))
+    } catch (err) {
+      if (!isAxiosError(err)) {
+        return new UnknownKratosError(err)
+      }
+
+      if (err.message === "Request failed with status code 400") {
+        // FIXME: not the right error. we expect the identity to exist
+        return new LikelyUserAlreadyExistError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+
+    if (identity.schema_id !== SchemaIdType.PhoneNoPasswordV0) {
+      return new IncompatibleSchemaUpgradeError()
+    }
+
+    if (identity.state === undefined)
+      throw new KratosError("state undefined, probably impossible state") // type issue
+
+    identity.traits = { ...identity.traits, email }
+
+    const adminIdentity: UpdateIdentityBody = {
+      ...identity,
+      credentials: { password: { config: { password } } },
+      state: identity.state,
+      schema_id: "phone_email_no_password_v0",
+    }
+
+    try {
+      const { data: newIdentity } = await kratosAdmin.updateIdentity({
+        id: kratosUserId,
+        updateIdentityBody: adminIdentity,
+      })
+
+      return toDomainIdentityEmailPhone(newIdentity)
+    } catch (err) {
+      return (
+        (isAxiosError(err) && new UnknownKratosError(err.message)) ||
+        new UnknownKratosError(err)
+      )
+    }
+  }
+
+  const removeEmailFromIdentity = async ({ kratosUserId }: { kratosUserId: UserId }) => {
+    let identity: KratosIdentity
+
+    try {
+      ;({ data: identity } = await kratosAdmin.getIdentity({ id: kratosUserId }))
+    } catch (err) {
+      return (
+        (isAxiosError(err) && new UnknownKratosError(err.message)) ||
+        new UnknownKratosError(err)
+      )
+    }
+
+    if (identity.schema_id !== SchemaIdType.PhoneEmailNoPasswordV0) {
+      return new IncompatibleSchemaUpgradeError()
+    }
+
+    if (identity.state === undefined)
+      throw new KratosError("state undefined, probably impossible state") // type issue
```

Comment:
This should be a specific error `KratodIdentityUndefined` or something like that.